File: demo_scala/src/main/java/com/vova/Demo_022_class_override.java
package com.vova;

/**
 * scala中属性和方法都是动态绑定，而java中只有方法为动态绑定
 * @author WangYang - vova
 * @version Create in 3:08 下午 2020/7/8
 */
class Person {

    public String name = "person";
    public void hello() {
        System.out.println("hello person");
    }

}
class Teacher extends Person {

    public String name = "teacher";
    @Override
    public void hello() {
        System.out.println("hello teacher");
    }

}

public class Demo_022_class_override {

    public static void main(String[] args) {
        Teacher teacher = new Teacher();
        Person person = new Teacher();

        System.out.println(teacher.name);
        teacher.hello();

        System.out.println(person.name);
        person.hello();
    }
}

File: demo_scala/src/main/java/com/vova/demo_022_class_override.scala
package com.vova

/**
 * scala中属性和方法都是动态绑定，而java中只有方法为动态绑定
 */
object demo_022_class_override {

    class Person {
        val name: String = "person"

        def hello(): Unit = {
            println("hello person")
        }
    }

    class Teacher extends Person {

        override val name: String = "teacher"

        override def hello(): Unit = {
            println("hello teacher")
        }
    }

    def main(args: Array[String]): Unit = {
        val teacher: Teacher = new Teacher()
        println(teacher.name)
        teacher.hello()

        val teacher1: Person = new Teacher
        println(teacher1.name)
        teacher1.hello()
    }
    

}
